### PR TITLE
Make HAMT attack resistant by using cryptographic hashing

### PR DESCRIPTION
**Pull request by krl**

### CODE REVIEW

File: src/sink.rs [hunk 1]
```diff
@@ -32,9 +33,7 @@ impl<'a, H: ByteHash> Sink<'a, H> {
     pub(crate) fn fin(self) -> io::Result<H::Digest> {
         let Sink { bytes, .. } = self;
         let mut hasher = H::state();
-        hasher
-            .write_all(&bytes)
-            .expect("In memory write should always succeed");
+        hasher.write(&bytes);
```
Comment (jsulmont) [inline src/sink.rs]: shouldn't we call `write_all`?

File: structures/hamt/src/lib.rs [hunk 2]
```diff
@@ -27,53 +26,51 @@ impl<K: Content<H>, V: Content<H>, H: ByteHash> Default for HAMT<K, V, H> {
     }
 }
 
-#[inline(always)]
-fn hash<T: Hash>(t: T) -> u64 {
-    let mut hasher = SeaHasher::new();
-    t.hash(&mut hasher);
-    hasher.finish()
+fn select_slot(hash: &[u8], depth: usize) -> usize {
```
Comment (jsulmont) [inline structures/hamt/src/lib.rs]: ok

File: src/sink.rs (hunk 1 above)
Comment (krl) [reply src/sink.rs]: This replaces io::Write with hasher::Hasher

**Comment by jsulmont** [comment]
Nice!

**Pull request merged**